runtime/src/main/java/uk/ac/manchester/tornado/runtime/TornadoVM.java

```diff
@@ -156,7 +158,9 @@ public TornadoVM(TornadoExecutionContext graphContext, byte[] code, int limit, T
             final int deviceIndex = buffer.getInt();
             debug("loading context %s", contexts.get(deviceIndex));
             final long t0 = System.nanoTime();
-            contexts.get(deviceIndex).ensureLoaded();
+            if (contexts.get(deviceIndex) != null) {
+                contexts.get(deviceIndex).ensureLoaded();
```

Comment:
Since now there is a mapping between <device, TornadoVM object>, is there any reason to keep a list of `contexts` inside each TornadoVM object?

Reply:
The mapping is between a logic device ID (abstract representation in TornaoVM) and the physical device. Yes, we need to keep this high-level mapping for the concurrent BC interpreter.